flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java

```diff
@@ -68,11 +67,20 @@
     @Before
     public void setUp() throws Exception {
         super.setUp();
-        final CompletedCheckpointStore completedCheckpointStore =
-                new EmbeddedCompletedCheckpointStore();
         haServices.setCheckpointRecoveryFactory(
-                PerJobCheckpointRecoveryFactory.useSameServicesForAllJobs(
-                        completedCheckpointStore, new StandaloneCheckpointIDCounter()));
+                new PerJobCheckpointRecoveryFactory<EmbeddedCompletedCheckpointStore>(
+                        (maxCheckpoints, previous) -> {
+                            if (previous != null) {
+                                // First job attempt failed before cleaning up the checkpoint store.
+                                // This means we need to use the actual checkpoint store state for
+                                // recovery (instead of suspended checkpoints as usual).
```

Comment:
Hmm, if we remove the `suspendedCheckpoints` field from the `EmbeddedCompletedCheckpointStore` and remove checkpoints from `checkpoints` on `shutDown(GloballyTerminalState)`, then we wouldn't need this extra logic here. The contract would then be: Whatever is returned by `store.getAllCheckpoints()` will be recovered.

Reply:
I think we should also expose shutdown status for some advanced assert (like this one), but this should be doable 🤔

Reply:
fixed